File: algos/rotate-matrix.ts
// rotate a number[][] matrix 90 degrees
// [1,2]  =>  [3, 1]
// [3,4]      [4, 2]

function rotate(matrix: number[][]): number[][] {
  let layer = 0;
  while (layer < Math.floor(matrix.length / 2)) {
    const first = layer;
    const last = matrix.length - 1 - layer;
    for (let i = first; i < last; i++) {
      const offset = i - layer; // this increases as i increases - going across the top row
      // top left matrix[first][i]
      // top right matrix[first + offset][last]
      // lower right matrix[last][last - offset]
      // lower left matrix[last - offset][first]

      // swap three times

      [matrix[first][i], matrix[first + offset][last]] = [
        matrix[first + offset][last],
        matrix[first][i],
      ];

      [matrix[first][i], matrix[last - offset][first]] = [
        matrix[last - offset][first],
        matrix[first][i],
      ];
      [matrix[last][last - offset], matrix[last - offset][first]] = [
        matrix[last - offset][first],
        matrix[last][last - offset],
      ];
    }
    layer++;
  }

  return matrix;
}

console.log(
  rotate([
    [1, 2],
    [3, 4],
  ]),
);

console.log(
  rotate([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
  ]),
);

console.log(
  rotate([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12],
    [13, 14, 15, 16],
  ]),
);
// console.log('rotate[[1]]', rotate([[1]]))
// console.log('rotate[[1, 2], [3,4]]', rotate([[1,2], [3,4]]))
// console.log('rotate[[1, 2, 3], [4, 5, 6], [7, 8, 9]]', rotate([[1, 2, 3], [4, 5, 6], [7, 8, 9]]))
// console.log('rotate[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 16]]', rotate([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 16]]))

File: algos/rotate_matrix.py
from types import List


class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        # 2 pointers for left and right, and have them cave in per iteration
        # don't overthink the grids. Use left and right as boundaries first, then do the offset math

        # use pointers to set outer boundaries. have them cave in per iteration
        start, end = 0, len(matrix[0]) - 1
        while start < end:
            # do inner rotation end - start times
            for i in range(end - start):
                # topLeft matrix[start][start]
                # topRight matrix[start][end]
                # bottomRight matrix[end][end]
                # bottomLeft matrix[end][start]

                # now, try to figure out offset logic with {i}
                # topLeft matrix[start][start + i]
                # topRight matrix[start + i][end]
                # bottomRight matrix[end][end - i]
                # bottomLeft matrix[end - i][start]

                # rotate three times
                matrix[start][start + i], matrix[start + i][end] = matrix[start + i][end], matrix[start][start + i]
                matrix[start][start + i], matrix[end][end - i] = matrix[end][end - i], matrix[start][start + i]
                matrix[start][start + i], matrix[end - i][start] = matrix[end - i][start], matrix[start][start + i]
                # print('matrix', matrix)
            # shrink box
            start += 1
            end -= 1
